File: app/src/main/java/in/woloo/www/app/WolooApplication.java
package in.woloo.www.app;

import static com.facebook.FacebookSdk.AUTO_LOG_APP_EVENTS_ENABLED_PROPERTY;
import static com.facebook.FacebookSdk.setAdvertiserIDCollectionEnabled;
import static com.facebook.FacebookSdk.setAutoLogAppEventsEnabled;

import android.annotation.SuppressLint;
import android.app.Application;
import android.content.Context;
import android.content.IntentFilter;
import android.os.Build;
import android.text.TextUtils;
import android.util.Log;

import in.woloo.www.common.CommonUtils;
import in.woloo.www.utils.Logger;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.appsflyer.AppsFlyerConversionListener;
import com.appsflyer.AppsFlyerLib;
import com.facebook.FacebookSdk;
import com.facebook.GraphRequest;
import com.facebook.GraphResponse;
import com.facebook.LoggingBehavior;
import com.facebook.appevents.AppEventsConstants;
import com.facebook.appevents.AppEventsLogger;
import com.facebook.applinks.AppLinkData;
import com.google.android.gms.maps.MapsInitializer;
import com.google.android.gms.maps.OnMapsSdkInitializedCallback;
import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.google.firebase.messaging.FirebaseMessaging;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.netcore.android.Smartech;
import com.netcore.android.logger.SMTDebugLevel;
import com.netcore.android.smartechpush.SmartPush;
import com.netcore.android.smartechpush.notification.SMTNotificationListener;
import com.netcore.android.smartechpush.notification.SMTNotificationOptions;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;

import dev.b3nedikt.restring.Restring;
import dev.b3nedikt.reword.RewordInterceptor;
import dev.b3nedikt.viewpump.ViewPump;
import in.woloo.www.BuildConfig;
import in.woloo.www.R;
import in.woloo.www.dashboard.ui.home.model.NearByStoreResponse;
import in.woloo.www.fcm.DeeplinkReceiver;
import in.woloo.www.more.models.UserProfileMergedResponse;
import in.woloo.www.more.models.ViewProfileResponse;
import in.woloo.www.search.SearchWolooResponse;
import in.woloo.www.utils.Logger;
import in.woloo.www.v2.profile.model.UserProfile;
import io.hansel.core.logger.HSLLogLevel;
import io.hansel.hanselsdk.Hansel;
import io.hansel.hanselsdk.HanselDeepLinkListener;
import io.hansel.ujmtracker.HanselInternalEventsListener;
import io.hansel.ujmtracker.HanselTracker;

public class WolooApplication extends Application implements SMTNotificationListener {

    private static WolooApplication mInstance;

    private static final String AF_DEV_KEY = "i6aoJbBqs6pWjzSvo5hbtJ";
    private NearByStoreResponse.Data nearByWoloo;
    private int likeStatus = -1;
    private UserProfile profileResponse;
    private ViewProfileResponse subscriptionViewProfileResponse;
    private SearchWolooResponse.Data.Woloo data;
    public String TAG = WolooApplication.class.getSimpleName();

    private static String errorMessage = "";

    /*calling on onCreate*/
    @SuppressLint("RestrictedApi")
    @Override
    public void onCreate() {
        super.onCreate();
        FacebookSdk.sdkInitialize(getApplicationContext());
        Log.d("FB SDK 1" , FacebookSdk.getApplicationId());
        //MapsInitializer.initialize(getApplicationContext(), MapsInitializer.Renderer.LATEST, this);
        /*AppLinkData.fetchDeferredAppLinkData(this,
                appLinkData -> {
                    Toast.makeText(this, appLinkData.toString(), Toast.LENGTH_SHORT).show();
                }
        );*/
        AppEventsLogger.activateApp(this);

        setAutoLogAppEventsEnabled(true); // Added by Aarati


        FacebookSdk.setAutoInitEnabled(true); // Added by Aarati

        FacebookSdk.fullyInitialize();// Added by Aarati

        setAdvertiserIDCollectionEnabled(true);

        AppEventsLogger logger = AppEventsLogger.newLogger(this);
        logger.logEvent(AppEventsConstants.EVENT_PARAM_PRODUCT_APPLINK_ANDROID_APP_NAME);
        FacebookSdk.setIsDebugEnabled(true);
        FacebookSdk.addLoggingBehavior(LoggingBehavior.APP_EVENTS);


        mInstance = this;
        setCrashlytics();

        AppsFlyerConversionListener conversionListener = new AppsFlyerConversionListener() {

            @Override
            public void onConversionDataSuccess(Map<String, Object> conversionData) {
                for (String attrName : conversionData.keySet()) {
                    Logger.d("LOG_TAG", "attribute: " + attrName + " = " + conversionData.get(attrName));
                }
            }

            @Override
            public void onConversionDataFail(String s) {

            }

            @Override
            public void onAppOpenAttribution(Map<String, String> conversionData) {
                for (String attrName : conversionData.keySet()) {
                    Logger.d("LOG_TAG", "attribute: " + attrName + " = " + conversionData.get(attrName));
                }
            }

            @Override
            public void onAttributionFailure(String s) {

            }
        };
        AppsFlyerLib.getInstance().init(AF_DEV_KEY, conversionListener, this);
        AppsFlyerLib.getInstance().start(this);

        Restring.init(this);
        ViewPump.init(RewordInterceptor.INSTANCE);
        Logger.i(TAG, "onCreate");
        setUpNetCore();
    }

    private void setUpNetCore(){
        Smartech smartech = Smartech.getInstance(new WeakReference<>(this.getApplicationContext()));
        smartech.initializeSdk(this);
        smartech.trackAppInstallUpdateBySmartech();

        if (BuildConfig.DEBUG) {
            smartech.setDebugLevel(SMTDebugLevel.Level.VERBOSE); //9
            // enabling logs for nudges prodcut
            HSLLogLevel.all.setEnabled(true);
            HSLLogLevel.mid.setEnabled(true);
            HSLLogLevel.debug.setEnabled(true);
        }

        try {
            SmartPush smartPush = SmartPush.getInstance(new WeakReference<>(this));
            smartPush.fetchAlreadyGeneratedTokenFromFCM();
//            smartPush.setDevicePushToken(FirebaseInstanceId.getInstance().getToken());
            Logger.e(TAG, "token "+ smartPush.getDevicePushToken());
        } catch (Exception e) {
            Logger.e(TAG, "Fetching FCM token failed.");
        }
        Logger.v("DeeplinkReceiver", "Before");
        DeeplinkReceiver deeplinkReceiver = new DeeplinkReceiver();
        IntentFilter filter = new IntentFilter("com.smartech.EVENT_PN_INBOX_CLICK");
        //this.registerReceiver(deeplinkReceiver, filter); // commented by Aarati while version upgrade. and following block added
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            getApplicationContext().registerReceiver(deeplinkReceiver, filter, RECEIVER_EXPORTED);
        }
        Logger.v("DeeplinkReceiver", "After.");

        SMTNotificationOptions options = new SMTNotificationOptions(this);
        options.setBrandLogo("logo"); //e.g.logo is sample name for brand logo
        options.setLargeIcon("icon_nofification");//e.g.ic_notification is sample name for large icon
        options.setSmallIcon("ic_notification_woloo"); //e.g.ic_action_play is sample name for icon
        options.setSmallIconTransparent("ic_notification_woloo"); //e.g.ic_action_play is sample name for transparent small icon
        options.setTransparentIconBgColor("#414042");
        options.setPlaceHolderIcon("ic_notification");//e.g.ic_notification is sample name for placeholder icon
        SmartPush.getInstance(new WeakReference(this)).setNotificationOptions(options);
    }

    public static synchronized WolooApplication getInstance() {
        return mInstance;
    }


    public void setNearByWoloo(NearByStoreResponse.Data nearByWoloo) {
        this.nearByWoloo = nearByWoloo;
    }

    public NearByStoreResponse.Data getNearByWoloo() {
        return nearByWoloo;
    }

    public int getUpdatedLikeStatus() {
        return likeStatus;
    }

    public void setUpdatedLikeStatus(int likeStatus) {
        this.likeStatus = likeStatus;
    }

    public UserProfile getProfileResponse() {
        return profileResponse;
    }

    public void setProfileResponse(UserProfile profileResponse) {
        this.profileResponse = profileResponse;
    }

    public static void setErrorMessage(String message){
        errorMessage = message;
    }

    public static String getErrorMessage(){
        return errorMessage;
    }

    public void setSearchedWoloo(SearchWolooResponse.Data.Woloo data) {
        this.data = data;
    }

    public SearchWolooResponse.Data.Woloo getSearchedWoloo() {
        return data;
    }

    /*calling on setCrashlytics*/
    private void setCrashlytics() {
        Logger.i(TAG, "setCrashlytics");
        try {
            String mStrLiveUrl = BuildConfig.LIVE_URL;
            Logger.e("MSTRLIVE", mStrLiveUrl);
            FirebaseCrashlytics crashlytics = FirebaseCrashlytics.getInstance();
            //in firebase there was a crash on this line (NullPointerException)
            if (mStrLiveUrl != null) {
                if (mStrLiveUrl.equalsIgnoreCase("3")) {
                    Logger.e("MSTRLIVE", "CHCK1");
                    crashlytics.setCrashlyticsCollectionEnabled(true);
                } else {
                    crashlytics.setCrashlyticsCollectionEnabled(false);
                }
            }
        } catch (Exception e) {
            Logger.e("APP EXCEPTION", e.getMessage());
              CommonUtils.printStackTrace(e);
        }

    }


    @Override
    public void getSmartechNotifications(@NonNull String data, int from) {
        SmartPush.getInstance(new WeakReference<Context>(getApplicationContext())).renderNotification(data,from);
    }
}

File: app/src/main/java/in/woloo/www/app/WolooApplication.kt
package `in`.woloo.www.app

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.IntentFilter
import android.os.Build
import android.util.Log
import com.appsflyer.AppsFlyerConversionListener
import com.appsflyer.AppsFlyerLib
import com.facebook.FacebookSdk.addLoggingBehavior
import com.facebook.FacebookSdk.fullyInitialize
import com.facebook.FacebookSdk.sdkInitialize
import com.facebook.FacebookSdk.setAdvertiserIDCollectionEnabled
import com.facebook.FacebookSdk.setAutoInitEnabled
import com.facebook.FacebookSdk.setAutoLogAppEventsEnabled
import com.facebook.FacebookSdk.setIsDebugEnabled
import com.facebook.LoggingBehavior
import com.facebook.appevents.AppEventsConstants.EVENT_PARAM_PRODUCT_APPLINK_ANDROID_APP_NAME
import com.facebook.appevents.AppEventsLogger
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.netcore.android.Smartech
import com.netcore.android.logger.SMTDebugLevel
import com.netcore.android.smartechpush.SmartPush
import com.netcore.android.smartechpush.notification.SMTNotificationListener
import com.netcore.android.smartechpush.notification.SMTNotificationOptions
import dev.b3nedikt.restring.Restring.init
import dev.b3nedikt.reword.RewordInterceptor
import dev.b3nedikt.viewpump.ViewPump.init
import `in`.woloo.www.BuildConfig
import `in`.woloo.www.application_kotlin.database.SharedPrefSettings
import `in`.woloo.www.application_kotlin.model.server_response.NearByStoreResponse
import `in`.woloo.www.application_kotlin.model.server_response.SearchWolooResponse
import `in`.woloo.www.common.CommonUtils
import `in`.woloo.www.application_kotlin.fcm.DeeplinkReceiver
import `in`.woloo.www.more.models.ViewProfileResponse
import `in`.woloo.www.utils.Logger.d
import `in`.woloo.www.utils.Logger.e
import `in`.woloo.www.utils.Logger.i
import `in`.woloo.www.utils.Logger.v
import `in`.woloo.www.v2.profile.model.UserProfile
import io.hansel.core.logger.HSLLogLevel
import java.lang.ref.WeakReference
import android.content.res.Configuration
import android.os.Bundle
import androidx.annotation.OptIn
import androidx.media3.common.util.UnstableApi
import androidx.media3.datasource.cache.LeastRecentlyUsedCacheEvictor
import androidx.media3.datasource.cache.SimpleCache
import com.facebook.FacebookSdk
import java.io.File

class WolooApplication : Application(), SMTNotificationListener {
    @JvmField
    var nearByWoloo: NearByStoreResponse.Data? = null
    var updatedLikeStatus: Int = -1
    var profileResponseKotlin: UserProfile? = null
    private val subscriptionViewProfileResponse: ViewProfileResponse? = null
    private var data: SearchWolooResponse.Data.Woloo? = null
    var TAG: String = WolooApplication::class.java.simpleName

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (newConfig.fontScale != 1.0f) {
            val config = Configuration(newConfig)
            config.fontScale = 1.0f
            resources.updateConfiguration(config, resources.displayMetrics)
        }
    }

    /*calling on onCreate*/
    @OptIn(UnstableApi::class)
    @SuppressLint("RestrictedApi")
    override fun onCreate() {
        super.onCreate()
        sdkInitialize(applicationContext)
        ExoPlayerCache.init(this)
        instance = this

        registerActivityLifecycleCallbacks(object : ActivityLifecycleCallbacks {
            override fun onActivityResumed(activity: Activity) {
                currentActivity = activity
            }
            override fun onActivityPaused(activity: Activity) {
                if (currentActivity === activity) currentActivity = null
            }
            override fun onActivityDestroyed(activity: Activity) {}
            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
            override fun onActivityStarted(activity: Activity) {}
            override fun onActivityStopped(activity: Activity) {}
            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
        })
        //
      //  Log.d("FB SDK 1", applicationId!!)
        //MapsInitializer.initialize(getApplicationContext(), MapsInitializer.Renderer.LATEST, this);
        /*AppLinkData.fetchDeferredAppLinkData(this,
                appLinkData -> {
                    Toast.makeText(this, appLinkData.toString(), Toast.LENGTH_SHORT).show();
                }
        );*/
        AppEventsLogger.activateApp(this)


        setAutoLogAppEventsEnabled(true) // Added by Aarati


        setAutoInitEnabled(true) // Added by Aarati

        fullyInitialize() // Added by Aarati

        setAdvertiserIDCollectionEnabled(true)

        val logger = AppEventsLogger.newLogger(this)
        logger.logEvent(EVENT_PARAM_PRODUCT_APPLINK_ANDROID_APP_NAME)
        setIsDebugEnabled(true)
        addLoggingBehavior(LoggingBehavior.APP_EVENTS)


        instance = this
        setCrashlytics()

        val conversionListener: AppsFlyerConversionListener = object : AppsFlyerConversionListener {
            override fun onConversionDataSuccess(conversionData: Map<String, Any>) {
                for (attrName in conversionData.keys) {
                    d("LOG_TAG", "attribute: " + attrName + " = " + conversionData[attrName])
                }
            }

            override fun onConversionDataFail(s: String) {
            }

            override fun onAppOpenAttribution(conversionData: Map<String, String>) {
                for (attrName in conversionData.keys) {
                    d("LOG_TAG", "attribute: " + attrName + " = " + conversionData[attrName])
                }
            }

            override fun onAttributionFailure(s: String) {
            }
        }
        AppsFlyerLib.getInstance().init(
            AF_DEV_KEY, conversionListener,
            this
        )
        AppsFlyerLib.getInstance().start(this)

        logCustomAppInstall()

        init(this)
        init(RewordInterceptor)
        i(TAG, "onCreate")
        setUpNetCore()

    }

    private fun setUpNetCore() {
        val smartech = Smartech.getInstance(WeakReference(this.applicationContext))
        smartech.initializeSdk(this)
        smartech.trackAppInstallUpdateBySmartech()

        if (BuildConfig.DEBUG) {
            smartech.setDebugLevel(SMTDebugLevel.Level.VERBOSE) //9
            // enabling logs for nudges prodcut
            HSLLogLevel.all.isEnabled = true
            HSLLogLevel.mid.isEnabled = true
            HSLLogLevel.debug.isEnabled = true
        }

        try {
            val smartPush = SmartPush.getInstance(WeakReference(this))
            smartPush.fetchAlreadyGeneratedTokenFromFCM()
            //            smartPush.setDevicePushToken(FirebaseInstanceId.getInstance().getToken());
            e(TAG, "token " + smartPush.getDevicePushToken())
        } catch (e: Exception) {
            e(TAG, "Fetching FCM token failed.")
        }
        v("DeeplinkReceiver", "Before")
        val deeplinkReceiver = DeeplinkReceiver()
        val filter = IntentFilter("com.smartech.EVENT_PN_INBOX_CLICK")
        //this.registerReceiver(deeplinkReceiver, filter); // commented by Aarati while version upgrade. and following block added
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            applicationContext.registerReceiver(deeplinkReceiver, filter, RECEIVER_EXPORTED)
        }
        v("DeeplinkReceiver", "After.")

        val options = SMTNotificationOptions(this)
        options.brandLogo = "logo" //e.g.logo is sample name for brand logo
        options.largeIcon = "icon_nofification" //e.g.ic_notification is sample name for large icon
        options.smallIcon = "ic_notification_woloo" //e.g.ic_action_play is sample name for icon
        options.smallIconTransparent =
            "ic_notification_woloo" //e.g.ic_action_play is sample name for transparent small icon
        options.transparentIconBgColor = "#414042"
        options.placeHolderIcon =
            "ic_notification" //e.g.ic_notification is sample name for placeholder icon
        SmartPush.getInstance(WeakReference(this)).setNotificationOptions(options)
    }

    fun getProfileResponse(): UserProfile? {
        return profileResponseKotlin
    }

    fun setProfileResponse(profileResponseKotlin: UserProfile?) {
        this.profileResponseKotlin = profileResponseKotlin
    }

    var searchedWoloo: SearchWolooResponse.Data.Woloo?
        get() = data
        set(data) {
            this.data = data
        }

    /*calling on setCrashlytics*/
    private fun setCrashlytics() {
        i(TAG, "setCrashlytics")
        try {
            val mStrLiveUrl = BuildConfig.LIVE_URL
            e("MSTRLIVE", mStrLiveUrl)
            val crashlytics = FirebaseCrashlytics.getInstance()
            //in firebase there was a crash on this line (NullPointerException)
            if (mStrLiveUrl != null) {
                if (mStrLiveUrl.equals("3", ignoreCase = true)) {
                    e("MSTRLIVE", "CHCK1")
                    crashlytics.isCrashlyticsCollectionEnabled = true
                } else {
                    crashlytics.isCrashlyticsCollectionEnabled = false
                }
            }
        } catch (e: Exception) {
            e("APP EXCEPTION", e.message!!)
            CommonUtils.printStackTrace(e)
        }
    }


    @UnstableApi
    object ExoPlayerCache {

        private const val MAX_CACHE_SIZE = 100L * 1024 * 1024 // 100 MB

        private var _simpleCache: SimpleCache? = null

        fun init(context: Context) {
            if (_simpleCache == null) {
                val cacheDir = File(context.cacheDir, "exo_cache")
                _simpleCache = SimpleCache(cacheDir, LeastRecentlyUsedCacheEvictor(MAX_CACHE_SIZE))
            }
        }

        val simpleCache: SimpleCache
            get() = _simpleCache ?: throw IllegalStateException("ExoPlayerCache not initialized")
    }



    override fun getSmartechNotifications(data: String, from: Int) {
        SmartPush.getInstance(
            WeakReference(
                applicationContext
            )
        ).renderNotification(data, from)
    }

    private fun logCustomAppInstall() {

    }

    companion object {
        @get:Synchronized
        var instance: WolooApplication? = null
            private set

        private const val AF_DEV_KEY = "i6aoJbBqs6pWjzSvo5hbtJ"
        @JvmField
        var errorMessage: String = ""

        var currentActivity: Activity? = null
    }
}
